File: packages/ssh-native/ios/HappierSshNativeBridge.swift
import ExpoModulesCore
import Foundation

@_silgen_name("happier_ssh_native_exec_json")
private func happierSshNativeExecJson(_ requestJson: UnsafePointer<CChar>) -> UnsafeMutablePointer<CChar>?

@_silgen_name("happier_ssh_native_start_loopback_tunnel_json")
private func happierSshNativeStartLoopbackTunnelJson(_ requestJson: UnsafePointer<CChar>) -> UnsafeMutablePointer<CChar>?

@_silgen_name("happier_ssh_native_stop_loopback_tunnel_json")
private func happierSshNativeStopLoopbackTunnelJson(_ tunnelId: UnsafePointer<CChar>) -> UnsafeMutablePointer<CChar>?

@_silgen_name("happier_ssh_native_cancel_request_json")
private func happierSshNativeCancelRequestJson(_ requestId: UnsafePointer<CChar>) -> UnsafeMutablePointer<CChar>?

@_silgen_name("happier_ssh_native_free_string")
private func happierSshNativeFreeString(_ value: UnsafeMutablePointer<CChar>?)

public enum HappierSshNativeBridge {
  private static let moduleVersion = "0.0.0"
  private static let rustWorkQueue = DispatchQueue(
    label: "dev.happier.ssh-native.rust-work",
    qos: .userInitiated,
    attributes: .concurrent
  )
  fileprivate static let promptLock = NSLock()
  fileprivate static var promptResponses: [String: HostKeyPromptWaiter] = [:]

  public static func availability() -> [String: Any] {
    [
      "available": true,
      "platform": "ios",
      "engine": "russh",
      "moduleVersion": moduleVersion,
      "supportsLoopbackTunnel": true,
      "supportsPersistentHostKeyStorage": false,
    ]
  }

  public static func respondToHostKeyPrompt(promptId: String, response: [String: Any]) {
    promptLock.lock()
    let waiter = promptResponses.removeValue(forKey: promptId)
    promptLock.unlock()
    waiter?.complete(response)
  }

  public static func respondToAuthPrompt(promptId: String, response: [String: Any]) {
    promptLock.lock()
    let waiter = promptResponses.removeValue(forKey: promptId)
    promptLock.unlock()
    waiter?.complete(response)
  }

  public static func cancelRequest(requestId: String) {
    let trimmedRequestId = requestId.trimmingCharacters(in: .whitespacesAndNewlines)
    let cancellablePromptIds = Set([
      "host-key-\(trimmedRequestId)",
      "auth-passphrase-\(trimmedRequestId)",
      "auth-kbi-\(trimmedRequestId)",
    ])
    promptLock.lock()
    let matchingPromptIds = promptResponses.keys.filter { cancellablePromptIds.contains($0) }
    let waiters = matchingPromptIds.compactMap { promptResponses.removeValue(forKey: $0) }
    promptLock.unlock()
    for waiter in waiters {
      waiter.complete([
        "decision": "reject",
        "reason": "Native SSH request was cancelled.",
      ])
    }
    requestId.withCString { pointer in
      if let response = happierSshNativeCancelRequestJson(pointer) {
        happierSshNativeFreeString(response)
      }
    }
  }

  public static func exec(module: Module, request: [String: Any]) throws -> [String: Any] {
    try withProgress(module: module, request: request, phase: "connecting") {
      try callRustJsonWithPrompts(module: module, request: request, happierSshNativeExecJson)
    }
  }

  public static func execAsync(module: Module, request: [String: Any]) async throws -> [String: Any] {
    try await runBlockingRustWork {
      try exec(module: module, request: request)
    }
  }

  public static func startLoopbackTunnel(module: Module, request: [String: Any]) throws -> [String: Any] {
    try withProgress(module: module, request: request, phase: "connecting") {
      try callRustJsonWithPrompts(module: module, request: request, happierSshNativeStartLoopbackTunnelJson)
    }
  }

  public static func startLoopbackTunnelAsync(module: Module, request: [String: Any]) async throws -> [String: Any] {
    try await runBlockingRustWork {
      try startLoopbackTunnel(module: module, request: request)
    }
  }

  public static func stopLoopbackTunnel(nativeTunnelId: String) {
    nativeTunnelId.withCString { pointer in
      if let response = happierSshNativeStopLoopbackTunnelJson(pointer) {
        happierSshNativeFreeString(response)
      }
    }
  }

  private static func runBlockingRustWork<T>(_ operation: @escaping () throws -> T) async throws -> T {
    try await withCheckedThrowingContinuation { continuation in
      rustWorkQueue.async {
        do {
          continuation.resume(returning: try operation())
        } catch {
          continuation.resume(throwing: error)
        }
      }
    }
  }
}

private func withProgress<T>(
  module: Module,
  request: [String: Any],
  phase: String,
  operation: () throws -> T
) throws -> T {
  sendProgress(module: module, request: request, phase: phase)
  defer {
    sendProgress(module: module, request: request, phase: "closing")
  }
  return try operation()
}

private func callRustJsonWithPrompts(
  module: Module,
  request: [String: Any],
  _ function: (UnsafePointer<CChar>) -> UnsafeMutablePointer<CChar>?
) throws -> [String: Any] {
  var currentRequest = request
  for _ in 0..<8 {
    let response = try callRustEnvelope(currentRequest, function)
    if response["ok"] as? Bool == true {
      guard let result = response["result"] as? [String: Any] else {
        throw nativeSshError(code: "engine-internal", message: "Native SSH engine returned an invalid result.")
      }
      return result
    }

    guard let error = response["error"] as? [String: Any] else {
      throw nativeSshError(code: "engine-internal", message: "Native SSH engine failed.")
    }
    let code = error["code"] as? String ?? "engine-internal"
    if (code == "host-key-untrusted" || code == "host-key-mismatch"),
       let prompt = error["hostKeyPrompt"] as? [String: Any] {
      let promptResponse = try waitForPromptResponse(
        module: module,
        request: currentRequest,
        eventName: "hostKeyPrompt",
        prompt: prompt,
        phase: "verifying-host-key",
        timeoutCode: "host-key-untrusted",
        timeoutMessage: "SSH host key trust prompt timed out."
      )
      if promptResponse["decision"] as? String == "reject" {
        throw nativeSshError(code: "host-key-rejected", message: "SSH host key trust was declined.")
      }
      currentRequest["hostKeyVerification"] = promptResponse
      continue
    }
    if code == "auth-prompt-required",
       let prompt = error["authPrompt"] as? [String: Any] {
      let promptResponse = try waitForPromptResponse(
        module: module,
        request: currentRequest,
        eventName: "authPrompt",
        prompt: prompt,
        phase: "authenticating",
        timeoutCode: "authentication-failed",
        timeoutMessage: "SSH authentication prompt timed out."
      )
      let decision = promptResponse["decision"] as? String
      if decision == "cancel" || decision == "reject" {
        throw nativeSshError(code: "auth-prompt-cancelled", message: "SSH authentication prompt was cancelled.")
      }
      currentRequest = retryRequestWithAuthPromptResponse(currentRequest, promptResponse)
      continue
    }
    throw nativeSshError(
      code: code,
      message: error["message"] as? String ?? "Native SSH engine failed."
    )
  }
  throw nativeSshError(code: "authentication-failed", message: "SSH authentication required too many prompt rounds.")
}

private func waitForPromptResponse(
  module: Module,
  request: [String: Any],
  eventName: String,
  prompt: [String: Any],
  phase: String,
  timeoutCode: String,
  timeoutMessage: String
) throws -> [String: Any] {
  guard let promptId = prompt["promptId"] as? String else {
    throw nativeSshError(code: "engine-internal", message: "Native SSH engine returned an invalid prompt.")
  }
  let waiter = HostKeyPromptWaiter()
  HappierSshNativeBridge.promptLock.lock()
  HappierSshNativeBridge.promptResponses[promptId] = waiter
  HappierSshNativeBridge.promptLock.unlock()
  sendProgress(module: module, request: request, phase: phase)
  module.sendEvent(eventName, prompt)
  let authTimeoutMs = readPositiveInt(request, "authTimeoutMs", fallback: 15_000)
  guard let response = waiter.wait(timeoutMs: authTimeoutMs) else {
    HappierSshNativeBridge.promptLock.lock()
    HappierSshNativeBridge.promptResponses.removeValue(forKey: promptId)
    HappierSshNativeBridge.promptLock.unlock()
    throw nativeSshError(code: timeoutCode, message: timeoutMessage)
  }
  return response
}

private func retryRequestWithAuthPromptResponse(
  _ request: [String: Any],
  _ response: [String: Any]
) -> [String: Any] {
  var retryRequest = request
  var auth = request["auth"] as? [String: Any] ?? [:]
  if let value = response["value"] as? String {
    auth["privateKeyPassphrase"] = value
    let previousAttempts = (auth["privateKeyPassphraseAttempts"] as? NSNumber)?.intValue
      ?? auth["privateKeyPassphraseAttempts"] as? Int
      ?? 0
    auth["privateKeyPassphraseAttempts"] = previousAttempts + 1
  }
  if let answers = response["answers"] as? [[String: Any]] {
    let existingAnswers = auth["keyboardInteractiveAnswers"] as? [String] ?? []
    let newAnswers = answers.compactMap { $0["value"] as? String }
    auth["keyboardInteractiveAnswers"] = existingAnswers + newAnswers
  }
  retryRequest["auth"] = auth
  return retryRequest
}

private func sendProgress(module: Module, request: [String: Any], phase: String) {
  module.sendEvent("progress", [
    "requestId": readString(request, "requestId"),
    "phase": phase,
    "host": readString(request, "host"),
    "port": readPositiveInt(request, "port", fallback: 22),
  ])
}

private func callRustJson(
  _ request: [String: Any],
  _ function: (UnsafePointer<CChar>) -> UnsafeMutablePointer<CChar>?
) throws -> [String: Any] {
  let response = try callRustEnvelope(request, function)
  if response["ok"] as? Bool == true {
    guard let result = response["result"] as? [String: Any] else {
      throw nativeSshError(code: "engine-internal", message: "Native SSH engine returned an invalid result.")
    }
    return result
  }
  let error = response["error"] as? [String: Any]
  throw nativeSshError(
    code: error?["code"] as? String ?? "engine-internal",
    message: error?["message"] as? String ?? "Native SSH engine failed."
  )
}

private func callRustEnvelope(
  _ request: [String: Any],
  _ function: (UnsafePointer<CChar>) -> UnsafeMutablePointer<CChar>?
) throws -> [String: Any] {
  let requestJson = try jsonString(request)
  guard let responsePointer = requestJson.withCString({ function($0) }) else {
    throw nativeSshError(code: "engine-internal", message: "Native SSH engine returned no result.")
  }
  defer { happierSshNativeFreeString(responsePointer) }

  let responseJson = String(cString: responsePointer)
  guard
    let responseData = responseJson.data(using: .utf8),
    let response = try JSONSerialization.jsonObject(with: responseData) as? [String: Any]
  else {
    throw nativeSshError(code: "engine-internal", message: "Native SSH engine returned an invalid result.")
  }
  return response
}

private func jsonString(_ value: [String: Any]) throws -> String {
  let data = try JSONSerialization.data(withJSONObject: value, options: [])
  guard let text = String(data: data, encoding: .utf8) else {
    throw nativeSshError(code: "engine-internal", message: "Invalid native SSH request.")
  }
  return text
}

private final class HostKeyPromptWaiter {
  private let condition = NSCondition()
  private var response: [String: Any]?

  func complete(_ response: [String: Any]) {
    condition.lock()
    self.response = response
    condition.signal()
    condition.unlock()
  }

  func wait(timeoutMs: Int) -> [String: Any]? {
    let deadline = Date().addingTimeInterval(TimeInterval(timeoutMs) / 1_000)
    condition.lock()
    while response == nil && Date() < deadline {
      condition.wait(until: deadline)
    }
    let value = response
    condition.unlock()
    return value
  }
}

private func readPositiveInt(_ value: [String: Any], _ key: String, fallback: Int) -> Int {
  if let number = value[key] as? Int, number > 0 {
    return number
  }
  if let number = value[key] as? NSNumber, number.intValue > 0 {
    return number.intValue
  }
  if let text = value[key] as? String, let number = Int(text), number > 0 {
    return number
  }
  return fallback
}

private func readString(_ value: [String: Any], _ key: String) -> String {
  (value[key] as? String)?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
}

private func nativeSshError(code: String, message: String) -> NSError {
  NSError(
    domain: "HappierSshNative",
    code: 1,
    userInfo: [
      "code": code,
      "message": message,
      NSLocalizedDescriptionKey: message,
    ]
  )
}

File: packages/ssh-native/android/src/main/java/dev/happier/ssh/HappierSshNativeBridge.kt
package dev.happier.ssh

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.exception.CodedException
import org.json.JSONArray
import org.json.JSONObject
import java.util.concurrent.CompletableFuture
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.TimeUnit
import java.util.concurrent.TimeoutException

object HappierSshNativeBridge {
  private const val MODULE_VERSION = "0.0.0"
  private val promptResponses = ConcurrentHashMap<String, CompletableFuture<Map<String, Any?>>>()
  private val rustLibraryLoaded: Boolean by lazy {
    runCatching { System.loadLibrary("happier_ssh_native") }.isSuccess
  }

  fun availability(): Map<String, Any> {
    if (!rustLibraryLoaded) {
      return mapOf(
        "available" to false,
        "reason" to "engine-unavailable",
        "detail" to "HappierSshNative russh core is not linked in this build."
      )
    }
    return mapOf(
      "available" to true,
      "platform" to "android",
      "engine" to "russh",
      "moduleVersion" to MODULE_VERSION,
      "supportsLoopbackTunnel" to true,
      "supportsPersistentHostKeyStorage" to false
    )
  }

  fun respondToHostKeyPrompt(promptId: String, response: Map<String, Any?>) {
    promptResponses.remove(promptId)?.complete(response)
  }

  fun respondToAuthPrompt(promptId: String, response: Map<String, Any?>) {
    promptResponses.remove(promptId)?.complete(response)
  }

  fun cancelRequest(requestId: String) {
    val trimmedRequestId = requestId.trim()
    val cancellablePromptIds = setOf(
      "host-key-$trimmedRequestId",
      "auth-passphrase-$trimmedRequestId",
      "auth-kbi-$trimmedRequestId"
    )
    promptResponses.entries
      .filter { cancellablePromptIds.contains(it.key) }
      .forEach { entry ->
        promptResponses.remove(entry.key)?.complete(mapOf(
          "decision" to "reject",
          "reason" to "Native SSH request was cancelled."
        ))
      }
    if (rustLibraryLoaded) {
      HappierSshNativeRust.cancelRequestJson(requestId)
    }
  }

  fun exec(module: Module, request: Map<String, Any?>): Map<String, Any?> {
    if (!rustLibraryLoaded) {
      throw NativeSshException("engine-unavailable", "HappierSshNative russh core is not linked in this build.")
    }
    return withProgress(module, request, "connecting") {
      val response = callRustWithPrompts(module, request) { requestJson ->
        HappierSshNativeRust.execJson(requestJson)
      }
      jsonObjectToMap(response)
    }
  }

  fun startLoopbackTunnel(module: Module, request: Map<String, Any?>): Map<String, Any?> {
    if (!rustLibraryLoaded) {
      throw NativeSshException("engine-unavailable", "HappierSshNative russh core is not linked in this build.")
    }
    return withProgress(module, request, "connecting") {
      val response = callRustWithPrompts(module, request) { requestJson ->
        HappierSshNativeRust.startLoopbackTunnelJson(requestJson)
      }
      jsonObjectToMap(response)
    }
  }

  private fun callRustWithPrompts(
    module: Module,
    request: Map<String, Any?>,
    operation: (String) -> String,
  ): JSONObject {
    var currentRequest = request
    repeat(8) {
      val response = callRust(currentRequest, operation)
      if (response.optBoolean("ok", false)) {
        return response.getJSONObject("result")
      }
      val error = response.optJSONObject("error")
      val code = error?.optString("code") ?: "engine-internal"
      val hostKeyPrompt = error?.optJSONObject("hostKeyPrompt")
      val authPrompt = error?.optJSONObject("authPrompt")
      if ((code == "host-key-untrusted" || code == "host-key-mismatch") && hostKeyPrompt != null) {
        val promptResponse = waitForPromptResponse(
          module,
          currentRequest,
          "hostKeyPrompt",
          hostKeyPrompt,
          "verifying-host-key",
          "host-key-untrusted",
          "SSH host key trust prompt timed out."
        )
        if (promptResponse["decision"] == "reject") {
          throw NativeSshException("host-key-rejected", "SSH host key trust was declined.")
        }
        currentRequest = currentRequest.toMutableMap().also {
          it["hostKeyVerification"] = promptResponse
        }
        return@repeat
      }
      if (code == "auth-prompt-required" && authPrompt != null) {
        val promptResponse = waitForPromptResponse(
          module,
          currentRequest,
          "authPrompt",
          authPrompt,
          "authenticating",
          "authentication-failed",
          "SSH authentication prompt timed out."
        )
        if (promptResponse["decision"] == "cancel" || promptResponse["decision"] == "reject") {
          throw NativeSshException("auth-prompt-cancelled", "SSH authentication prompt was cancelled.")
        }
        currentRequest = retryRequestWithAuthPromptResponse(currentRequest, promptResponse)
        return@repeat
      }
      throw NativeSshException(
        code,
        error?.optString("message") ?: "Native SSH engine failed."
      )
    }
    throw NativeSshException("authentication-failed", "SSH authentication required too many prompt rounds.")
  }

  private fun waitForPromptResponse(
    module: Module,
    request: Map<String, Any?>,
    eventName: String,
    prompt: JSONObject,
    phase: String,
    timeoutCode: String,
    timeoutMessage: String,
  ): Map<String, Any?> {
    val promptId = prompt.getString("promptId")
    val future = CompletableFuture<Map<String, Any?>>()
    promptResponses[promptId] = future
    sendProgress(module, request, phase)
    module.sendEvent(eventName, jsonObjectToMap(prompt))
    return try {
      future.get(readPositiveLong(request, "authTimeoutMs", 15_000L), TimeUnit.MILLISECONDS)
    } catch (_: TimeoutException) {
      throw NativeSshException(timeoutCode, timeoutMessage)
    } finally {
      promptResponses.remove(promptId)
    }
  }

  private fun retryRequestWithAuthPromptResponse(
    request: Map<String, Any?>,
    response: Map<String, Any?>,
  ): Map<String, Any?> {
    val retryRequest = request.toMutableMap()
    val auth = mutableMapOf<String, Any?>()
    (request["auth"] as? Map<*, *>)?.forEach { entry ->
      auth[entry.key.toString()] = entry.value
    }
    val value = response["value"]
    if (value is String) {
      auth["privateKeyPassphrase"] = value
      val previousAttempts = (auth["privateKeyPassphraseAttempts"] as? Number)?.toInt() ?: 0
      auth["privateKeyPassphraseAttempts"] = previousAttempts + 1
    }
    val promptAnswers = response["answers"]
    if (promptAnswers is Iterable<*>) {
      val existingAnswers = (auth["keyboardInteractiveAnswers"] as? Iterable<*>)
        ?.mapNotNull { it as? String }
        ?: emptyList()
      val newAnswers = promptAnswers.mapNotNull { entry ->
        (entry as? Map<*, *>)?.get("value") as? String
      }
      auth["keyboardInteractiveAnswers"] = existingAnswers + newAnswers
    }
    retryRequest["auth"] = auth
    return retryRequest
  }

  private fun callRust(request: Map<String, Any?>, operation: (String) -> String): JSONObject {
    return JSONObject(operation(toJsonValue(request).toString()))
  }

  fun stopLoopbackTunnel(nativeTunnelId: String) {
    if (rustLibraryLoaded) {
      HappierSshNativeRust.stopLoopbackTunnelJson(nativeTunnelId)
    }
  }

  private fun <T> withProgress(
    module: Module,
    request: Map<String, Any?>,
    phase: String,
    operation: () -> T,
  ): T {
    sendProgress(module, request, phase)
    try {
      return operation()
    } finally {
      sendProgress(module, request, "closing")
    }
  }

  private fun sendProgress(module: Module, request: Map<String, Any?>, phase: String) {
    module.sendEvent("progress", mapOf(
      "requestId" to readString(request, "requestId"),
      "phase" to phase,
      "host" to readString(request, "host"),
      "port" to readPositiveInt(request, "port", 22),
    ))
  }
}

object HappierSshNativeRust {
  external fun execJson(requestJson: String): String
  external fun startLoopbackTunnelJson(requestJson: String): String
  external fun stopLoopbackTunnelJson(nativeTunnelId: String): String
  external fun cancelRequestJson(requestId: String): String
}

class NativeSshException(
  val nativeSshCode: String,
  message: String
) : CodedException(nativeSshCode, message, null)

private fun toJsonValue(value: Any?): Any {
  return when (value) {
    null -> JSONObject.NULL
    is Map<*, *> -> {
      val json = JSONObject()
      value.forEach { entry -> json.put(entry.key.toString(), toJsonValue(entry.value)) }
      json
    }
    is Iterable<*> -> {
      val json = JSONArray()
      value.forEach { item -> json.put(toJsonValue(item)) }
      json
    }
    is Array<*> -> {
      val json = JSONArray()
      value.forEach { item -> json.put(toJsonValue(item)) }
      json
    }
    else -> value
  }
}

private fun jsonObjectToMap(json: JSONObject): Map<String, Any?> {
  val result = mutableMapOf<String, Any?>()
  val keys = json.keys()
  while (keys.hasNext()) {
    val key = keys.next()
    result[key] = fromJsonValue(json.get(key))
  }
  return result
}

private fun fromJsonValue(value: Any?): Any? {
  return when (value) {
    null, JSONObject.NULL -> null
    is JSONObject -> jsonObjectToMap(value)
    is JSONArray -> (0 until value.length()).map { index -> fromJsonValue(value.get(index)) }
    else -> value
  }
}

private fun readPositiveLong(value: Map<String, Any?>, key: String, fallback: Long): Long {
  val raw = value[key]
  return when (raw) {
    is Number -> raw.toLong()
    is String -> raw.toLongOrNull()
    else -> null
  }?.takeIf { it > 0 } ?: fallback
}

private fun readPositiveInt(value: Map<String, Any?>, key: String, fallback: Int): Int {
  val raw = value[key]
  return when (raw) {
    is Number -> raw.toInt()
    is String -> raw.toIntOrNull()
    else -> null
  }?.takeIf { it > 0 && it <= 65_535 } ?: fallback
}

private fun readString(value: Map<String, Any?>, key: String): String {
  return (value[key] as? String)?.trim()?.takeIf { it.isNotEmpty() } ?: ""
}
